File: Single-diode model/Auxiliary Algorithms/Python/compute_5_parameters.py
import numpy as np
from scipy.optimize import least_squares
import matplotlib.pyplot as plt
import warnings

# --- Auxiliary Functions ---

def residuals_2_20(x, Voc, Isc, Vmp, Imp, q, k, Tref):
    """
    Calculates the residuals for the five-point model equations.
    This function is used by least_squares to find the model parameters.
    """
    Iph, Is0, A, Rs, Rp = x

    C = q / (A * k * Tref)
    cap = 700  # Safety cap to prevent overflow in exp()

    # Arguments for the exponential function, capped for stability
    a_sc = min(C*Rs*Isc, cap)
    a_oc = min(C*Voc, cap)
    a_mp = min(C*(Rs*Imp+Vmp), cap)
    a_eq5 = min(C*Is0, cap)

    F = np.zeros(5)
    F[0] = Isc - ( Iph - Is0*(np.exp(a_sc)-1) - (Rs*Isc)/Rp )
    F[1] = 0   - ( Iph - Is0*(np.exp(a_oc)-1) - (Voc)/Rp )
    F[2] = Imp - ( Iph - Is0*(np.exp(a_mp)-1) - (Rs*Imp+Vmp)/Rp )
    F[3] = Rs + (q*Is0*Rp*(Rs-Rp)/(A*k*Tref))*np.exp(a_eq5)
    term_coeff = 1 + q*(Vmp - Rs*Imp)/(A*k*Tref)
    F[4] = Iph - 2*Vmp/Rp + Is0 - Is0 * term_coeff * np.exp(a_mp)
      
    return F

# --- Main Script ---

## Module Parameters
ns = 60       # Number of series cells
n_p = 1       # Number of parallel branches 

Vmp_mod_ref = 30.1 # Voltage at maximum power point of the module (V)
Imp_mod_ref = 8.30 # Current at maximum power point of the module (A)
Voc_mod_ref = 37.2 # Open-circuit voltage of the module (V)
Isc_mod_ref = 8.87 # Short-circuit current of the module (A)

Tref = 25 + 273.15  # Reference temperature (K)
Sref = 1000         # Reference irradiance (W/m²)
alpha = 0.00065     # temperature coefficient of Isc (%/K)

## Operating Conditions
T = 44.5 + 273.15   # Current temperature (K)
S = 765           # Current irradiance (W/m²)

## PV cell parameters 
Vmp_cell_ref = Vmp_mod_ref / ns
Imp_cell_ref = Imp_mod_ref / n_p 
Voc_cell_ref = Voc_mod_ref / ns
Isc_cell_ref = Isc_mod_ref / n_p 

## Physical Constants
q = 1.60217662e-19  # Elementary charge (C)
k = 1.38064852e-23  # Boltzmann constant (J/K)
E_G0 = 1.166          # Band gap energy at 0K (eV)
k1 = 4.73e-4          # Coefficient k1 (eV/K)
k2 = 636              # Coefficient k2 (K)

## 4) Parameter Estimation via least_squares (more robust than fsolve)
# Initial guess based on MATLAB's x0 for better convergence
x0 = np.array([Isc_cell_ref, 1e-9, 3.0, 0.01, 2.0])

# Define physical bounds for the parameters ([Iph, Is0, A, Rs, Rp])
lb = [0, 0, 0.5, 0, 1e-3]  # Lower bounds
ub = [Isc_cell_ref * 1.2, 1e-6, 5.0, 1.0, 10000] # Upper bounds

result = least_squares(
    residuals_2_20, 
    x0, 
    args=(Voc_cell_ref, Isc_cell_ref, Vmp_cell_ref, Imp_cell_ref, q, k, Tref), 
    bounds=(lb, ub),
    method='trf', # Trust Region Reflective algorithm, good for bounds
    ftol=1e-10,
    xtol=1e-10,
    gtol=1e-10
)

if not result.success:
    warnings.warn(f'least_squares did not converge. Status: {result.status}, Message: {result.message}')

Iph_ref, Is0_ref, A, Rs, Rp = result.x
print(f'Iph_ref = {Iph_ref:.6f} A')
print(f'Is0_ref = {Is0_ref:.2e} A')
print(f'A       = {A:.6f}')
print(f'Rs      = {Rs:.6f} Ohms')
print(f'Rp      = {Rp:.6f} Ohms')

File: Single-diode model/Auxiliary Algorithms/MATLAB/compute_5_parameters.m
clc; clear; close all; format long

%% Module Parameters

ns      = 60;       % number of series cells

Vmp_mod_ref  = 30.1;     % voltage at maximum power point (V)
Imp_mod_ref  = 8.30;     % current at maximum power point (A)
Voc_mod_ref  = 37.2;     % open-circuit voltage (V)
Isc_mod_ref  = 8.87;     % short-circuit current (A)

Tref = 25 + 273.15; % Reference temperature (K)
Sref = 1000; % Reference irradiance (W/m²)

alpha   = 0.00065;  % temperature coefficient of Isc (%/K)
beta = -0.0037; % temperature coefficient of Voc (%/K)

%% Operating Conditions

T = 44.5 + 273.15; % Current temperature (K)
S = 765; % Current irradiance (W/m²)

%% Physical Constants

q = 1.60217662e-19; % Elementary charge (C)
k = 1.38064852e-23; % Boltzmann constant (J/K)
E_G0 = 1.166;            % Band gap energy at 0K (eV)
k1   = 4.73e-4;          % Coefficient k1 (eV/K)
k2   = 636;              % Coefficient k2 (K)

%% Parameter Estimation via fsolve

Rs_0 = (Voc_mod_ref - Vmp_mod_ref)/Imp_mod_ref;
Rp_0 = (Vmp_mod_ref)/(Isc_mod_ref - Imp_mod_ref);

x0 = [Isc_mod_ref; log10(1e-9); ns; log10(Rs_0); Rp_0];
opts = optimoptions('fsolve', ...
    'Display', 'off', 'TolFun', 1e-10, 'TolX', 1e-10, 'MaxIter', 1000, 'MaxFunctionEvaluations', 2000);
fun = @(x) residuals_2_20(x, Voc_mod_ref, Isc_mod_ref, Vmp_mod_ref, Imp_mod_ref, q, k, Tref);
[xsol, ~, exitflag] = fsolve(fun, x0, opts);

if exitflag <= 0
    warning('fsolve did not converge (exitflag = %d)', exitflag);
end

Iph_ref = xsol(1);
Is0_ref = xsol(2);
A = xsol(3);
Rs = xsol(4);
Rp = xsol(5);

fprintf('Iph_ref = %.6f A\n', Iph_ref);
fprintf('Is0_ref = %.2e A\n', Is0_ref);
fprintf('A       = %.6f\n', A);
fprintf('Rs      = %.6f Ohms\n', Rs);
fprintf('Rp      = %.6f Ohms\n', Rp);

%% --- Auxiliary Functions ---

function F = residuals_2_20(x, Voc, Isc, Vmp, Imp, q, k, Tref)
    Iph = x(1); Is0 = x(2); A = x(3); Rs = x(4); Rp = x(5);
    C   = q/(A*k*Tref);
    cap = 700;
    a_sc = min(C*Rs*Isc, cap);
    a_oc = min(C*Voc, cap);
    a_mp = min(C*(Rs*Imp+Vmp), cap);
    a_eq5 = min(C*Isc, cap); 
    F = zeros(5,1);
    F(1) = Isc - ( Iph - Is0*(exp(a_sc)-1) - (Rs*Isc)/Rp );
    F(2) = 0   - ( Iph - Is0*(exp(a_oc)-1) - (Voc)/Rp );
    F(3) = Imp - ( Iph - Is0*(exp(a_mp)-1) - (Rs*Imp+Vmp)/Rp );
    F(4) = Rs + (q*Is0*Rp*(Rs-Rp)/(A*k*Tref))*exp(a_eq5);
    term_coeff = 1 + q*(Vmp - Rs*Imp)/(A*k*Tref);
    F(5) = Iph - 2*Vmp/Rp + Is0 - Is0 * term_coeff * exp(a_mp);
end

function I = solve_I_V_2_11(V, Iph_ref, Is0_ref, A, Rs, Rp, ...
    q, k, S, Sref, alpha, T, Tref, E_G0, k1, k2, ns) 
    
    % 1. Photocurrent (Iph) 
    Iph = Iph_ref * (S / Sref) * (1 + alpha * (T - Tref));
    
    % 2. Calculation of Saturation Current (Is)
    
    % 2a. Calculation of band gap energy in eV at temperature T: Eg(T)
    Eg_T_eV = E_G0 - (k1 * T^2) / (T + k2);
    
    % 2b. Calculation of the temperature difference term
    temp_diff = (1 / Tref) - (1 / T);
    
    % 2c. Calculation of the exponent 
    exponent_term = (ns*q / (A * k)) * Eg_T_eV * temp_diff; 
    
    % 2d. Final calculation of Is(T)
    Is = Is0_ref * (T / Tref)^3 * exp(exponent_term);
    
    % 3. Iterative Solver (Newton-Raphson)
    I = Iph; % Initial guess
    for it = 1:30
        V_diode = V + I * Rs;
        arg_exp = min(q * V_diode / (A * k * T), 700); 
        expo = exp(arg_exp);
        
        f = Iph - Is * (expo - 1) - V_diode / Rp - I;
        df = -Is * expo * (q * Rs / (A * k * T)) - Rs / Rp - 1;
        
        dI = -f / df;
        I = I + dI;
        
        if abs(dI) < 1e-6, break;end
    end
end